### PR TITLE
UI functionality to choose collapse strategy exit criteria

### PR DESCRIPTION
**Pull request by gowthamrao**
UI functionality only. Additional work is needed to make changes to cohort-constructor logic based on user input.

UI for era-fy 

https://github.com/OHDSI/WebAPI/pull/234
https://github.com/OHDSI/WebAPI/issues/217
https://github.com/OHDSI/Atlas/issues/414
https://github.com/OHDSI/Atlas/issues/375
https://github.com/OHDSI/Atlas/issues/407

@chrisknoll

### CODE REVIEW

File: js/modules/cohortbuilder/components/CohortExpressionEditorTemplate.html [hunk 1]
```diff
@@ -46,7 +46,7 @@
 					<col span="1" class="rule" />
 					<col span="1" class="delete" />
 				</colgroup>
-				<tbody data-bind="sortable: {data: expression().PrimaryCriteria().CriteriaList, connectClass : 'primaryCriteria', options: {cancel: ':input, button, [contenteditable], .undraggable'}}">
+				<tbody data-bind="sortable: {data: expression().PrimaryCriteria().CriteriaList, connectClass : 'primaryCriteria', options: {cancel: ':input, button, [contenteditable]'}}">
```
Comment (chrisknoll) [inline js/modules/cohortbuilder/components/CohortExpressionEditorTemplate.html]: Hi, the 'undraggable' class is used to keep certain elements from being dragged in the ui, but you removed it from the class.  Why?

File: js/modules/cohortbuilder/components/CohortExpressionEditor.js [hunk 2]
```diff
@@ -288,29 +288,31 @@ define(['knockout', 'jquery', '../options', '../CriteriaGroup', '../CriteriaType
...
 
 		self.copyInclusionRule = function (inclusionRule) {
 			var copiedRule = new InclusionRule(ko.toJS(inclusionRule), self.expression().ConceptSets);
 			copiedRule.name("Copy of: " + copiedRule.name());
 			self.expression().InclusionRules.push(copiedRule);
 			self.selectedInclusionRule(copiedRule);
-		};
+		}
+
+		self.addConceptSet = function (item) {}
```
Comment (chrisknoll) [inline js/modules/cohortbuilder/components/CohortExpressionEditor.js]: This function might be unnecessary, I noticed elsewhere in the code that there's an event handler for the .addconceptset selector, but I'm wondering if that selector should be removed vs. adding a empty addConceptSet function...was there an error you were receiving where you needed to add this event handler?

File: js/modules/cohortbuilder/CohortExpression.js [hunk 3]
```diff
@@ -26,6 +27,7 @@ define(function (require, exports) {
 		self.CensoringCriteria = ko.observableArray(data.CensoringCriteria && data.CensoringCriteria.map(function (criteria) {
 			return CriteriaTypes.GetCriteriaFromObject(criteria, self.ConceptSets);
 		}));
+		self.CollapseSettings = {CollapseType: ko.observable(data.CollapseSettings && data.CollapseSettings.CollapseType || "Era"), EraPad: ko.observable(data.CollapseSettings && data.CollapseSettings.EraPad || "0") }
```
Comment (chrisknoll) [inline js/modules/cohortbuilder/CohortExpression.js]: Should this be initialized as numeric?

File: js/modules/cohortbuilder/components/CohortExpressionEditorTemplate.html [hunk 4]
```diff
@@ -99,7 +99,8 @@
 		</div>
 		<div data-bind="eventListener: [
 			 { event: 'click', selector: '.copyInclusionRule', callback: copyInclusionRule},
-			 { event: 'click', selector: '.deleteInclusionRule', callback: deleteInclusionRule}]">
+			 { event: 'click', selector: '.deleteInclusionRule', callback: deleteInclusionRule},
+			 { event: 'click', selector: '.addConceptSet', callback: addConceptSet}]">
```
Comment (chrisknoll) [inline js/modules/cohortbuilder/components/CohortExpressionEditorTemplate.html]: This event handler for .addConceptSet selector: is it necessary?  The callback function doesn't actually do anything, so I'm wondering why this was added?

**Comment by gowthamrao** [comment]
@chrisknoll  there was an error on our part that caused those issues. Can you delete this PR and we resubmit a clean new one -- or, can you review the corrections to the original PR we made.

**Comment by chrisknoll** [comment]
Hi, your corrections are fine, when I do the merge we squash it all, so feel free to make additional commits to this branch and it'll be folded into the final PR.  What I do when I review the PR is look at the overall lines changed and such, so your latest commit made the number of lines changed look less, so let me review again and I'll let you know if I see any more problems.

**Comment by chrisknoll** [comment]
Ok, 2 things:
1: I think code needs to be added to the cohortdefinitionviewer module to provide a print-friendly output for the selected collapse value.

2: there is no 'collapse type: none' option on the backed. Only option right now is 'Era'.  We always collapse:

```java
  public String getCollapseConstructorQuery(CollapseSettings collapseSettings) {
		// default constructor is era constructor. as more collapse strategies are introduced, the query template and parameters need to be changed to match.
		String query = ERA_CONSTRUCTOR_TEMPLATE;

		query = StringUtils.replace(query, "@eraGroup", "person_id");
		query = StringUtils.replace(query, "@eraconstructorpad", Integer.toString(collapseSettings.eraPad));
		return query;
  }
```

I wouldn't change anything to support a 'none' option.  We always need to collapse so showing 'none' is not a valid case  The question is how many different ways can we think of to collapse?  Right now, we just have era. We don't have any other option, so we just should put in the UI that we collapse events to eras, and they can specify a gap size between events to chain them together.  So I'd modify the UI to remove the dropdown, and just say 'Specify era collapse gap size: ' or something like that. It defaults to zero, if they leave it blank, it's assumed to be 0d etc.  

3. The value being default in the javascript code is 'Era' but the value should be "ERA".  It's case sensitive, and is based on the enum in java:
```java
public enum CollapseType {
	@JsonProperty("ERA")
	ERA;
}
```

**Comment by gowthamrao** [comment]
@chrisknoll done!

File: js/modules/cohortbuilder/components/CohortExpressionEditorTemplate.html [hunk 5]
```diff
@@ -204,5 +204,20 @@
...
+	<div role="tabpanel" data-bind="css: { active: $component.expressionMode() == 'exit' || $component.expressionMode() == 'all'}" class="tab-pane">
+		<div class="heading">
+			<b>Cohort Collapse Strategy</b>
+			This strategy collapses events into eras of a continous period and eliminates the duplicate rows for the same subjects.  
+		</div>
+		<div style="padding: 5px 0px" data-bind="with: expression().CollapseSettings">
+			<ul>
+				<li>
+					Specify era collapse gap size: 
+					<select data-bind="options: $component.options.dayOptions, value: EraPad"></select> days
```
Comment (chrisknoll) [inline js/modules/cohortbuilder/components/CohortExpressionEditorTemplate.html]: The input type for a numeric input isn't a select, rather it's done this way:

```
<span contenteditable="true" class="numericInputField dropdown" data-bind="htmlValue: EraPad.numeric(), eventType: 'blur', ko_autocomplete: { value: EraPad, source: $component.options.dayOptions, minLength: 0, maxShowItems: 10, scroll: true }" />
```

Could you modify the <select> element to be the <span> element that's content editable?

**Comment by chrisknoll** [comment]
I'm going to combine the 2 PRs open for the circe UI changes into one PR, and do the final changes to get this merged in.

**Pull request merged**